### PR TITLE
Fix #45: Adds exit status for commands and invalid commands

### LINKED ISSUE #45: Define and implement the exit code / status for commands exectuion
Since the output of plugin could be consumed by other tooling, different exit statuses from the plugin commands and sub-commands need to be defined based on the command execution and their statuses.

This should help other tools or user know if the execution of command was successful without parsing the output of command.

### PR DESCRIPTION
**Pull request by navidshaikh**
Fix #45 
- Exit with status 1, if:
  - Wrong command given / Catchall for general errors
- Exit with status 3 if:
  - Machine is not running
- Exit with status 126, if
  - Service inside the box is not running and user has asked for its connection information Or Command invoked cannot execute

Reference: http://www.tldp.org/LDP/abs/html/exitcodes.html

### CODE REVIEW

**Comment by brgnepal** [comment]
+1 for adding argument to `print_help` to accept status.

LGTM

**Comment by LalatenduMohanty** [comment]
@navidshaikh here is the guideline for exit codes http://www.tldp.org/LDP/abs/html/exitcodes.html 
I think we should exit 1 or 126  when machine is not running
Let me know your thoughts.

**Comment by navidshaikh** [comment]
@LalatenduMohanty : 

> I think we should exit 1 or 126 when machine is not running
> Let me know your thoughts.

I think wrong command given is a general error, and we are catching it by exiting with 1.
However  "machine is not running" is something specific to plugin and use case.

**Comment by LalatenduMohanty** [comment]
@navidshaikh I dont have a strong counter argument as I am not sure what will be the right exit code here i.e. when OpenShift service is not running . So unless @bexelbie has some opinion , we are good to go.

**Comment by navidshaikh** [comment]
PR needs rebase and update with exit statuses for merged PRs.

File: lib/vagrant-service-manager/command.rb [hunk 1]
```diff
@@ -21,7 +21,7 @@ def exit_if_machine_not_running
   vagrant up
               eos
             @env.ui.error(message)
-            exit 1
+            exit 2
```
Comment (bexelbie) [inline lib/vagrant-service-manager/command.rb]: The comment in this function may need to be updated.

I believe this is not an exit 2.  It either needs to be 3 (unreserved) or 1, imho.  Running this command without the machine running feels like an impermissible operation error.

File: lib/vagrant-service-manager/command.rb [hunk 2]
```diff
@@ -41,16 +41,16 @@ def execute
                 # display information about all the providers inside ADB/CDK
                 self.print_all_provider_info
             else
-                self.print_help
+                self.print_help(1)
```
Comment (bexelbie) [inline lib/vagrant-service-manager/command.rb:65]: I believe help should exit with a 0.

see

```
ls;echo $?
ls --help;echo $?
```

File: lib/vagrant-service-manager/command.rb [hunk 3]
```diff
@@ -41,16 +41,16 @@ def execute
...
             end
         when "help"
             self.print_help
```
Comment (bexelbie) [inline lib/vagrant-service-manager/command.rb:71]: See line 44

File: lib/vagrant-service-manager/command.rb [hunk 4]
```diff
@@ -41,16 +41,16 @@ def execute
...
         else
-            self.print_help
+            self.print_help(1)
```
Comment (bexelbie) [inline lib/vagrant-service-manager/command.rb:73]: see line 44

File: lib/vagrant-service-manager/command.rb [hunk 5]
```diff
@@ -109,8 +110,8 @@ def execute_openshift_info
           openshift_console_url = "#{openshift_url}/console"
           self.print_openshift_info(openshift_url, openshift_console_url)
         else
-          @env.ui.info("# OpenShift service is not running in the vagrant box.")
-          exit 1
+          @env.ui.error("# OpenShift service is not running in the vagrant box.")
+          exit 2
```
Comment (bexelbie) [inline lib/vagrant-service-manager/command.rb]: I vote for 126 here.

**Comment by bexelbie** [comment]
I've added some comments inline with the script.

This PR cannot be merged without docs and a changelog update.

File: lib/vagrant-service-manager/command.rb (hunk 2 above)
Comment (navidshaikh) [inline lib/vagrant-service-manager/command.rb:65]: if user has entered wrong command which we do not support in sub-commands, it should be non zero exit status.

```
ls --bex ; echo $?
```

File: lib/vagrant-service-manager/command.rb (hunk 1 above)
Comment (navidshaikh) [inline lib/vagrant-service-manager/command.rb]: Good catch, will update the PR and description.

File: lib/vagrant-service-manager/command.rb [hunk 6]
```diff
@@ -21,7 +21,7 @@ def exit_if_machine_not_running
   vagrant up
               eos
             @env.ui.error(message)
-            exit 1
+            exit 3
```
Comment (navidshaikh) [inline lib/vagrant-service-manager/command.rb:24]: @bexelbie : Updated the exit code and comment.

File: lib/vagrant-service-manager/command.rb [hunk 7]
```diff
@@ -109,8 +110,8 @@ def execute_openshift_info
           openshift_console_url = "#{openshift_url}/console"
           self.print_openshift_info(openshift_url, openshift_console_url)
         else
-          @env.ui.info("# OpenShift service is not running in the vagrant box.")
-          exit 1
+          @env.ui.error("# OpenShift service is not running in the vagrant box.")
+          exit 126
```
Comment (navidshaikh) [inline lib/vagrant-service-manager/command.rb:143]: @bexelbie Updated the exit code and PR description.

File: lib/vagrant-service-manager/command.rb (hunk 2 above)
Comment (bexelbie) [inline lib/vagrant-service-manager/command.rb:65]: Ok, I understand the exit code of 1 here.

Also, please stop making fun of my favorite ls option :P

File: lib/vagrant-service-manager/command.rb (hunk 3 above)
Comment (bexelbie) [inline lib/vagrant-service-manager/command.rb:71]: For clarity here should we put self.print_help(0) ?

Comment (navidshaikh) [inline lib/vagrant-service-manager/command.rb:71]: Since user has given wrong command, it will exit with 1 and show the help.

**Comment by navidshaikh** [comment]
Re-based and updated the PR with correct exit codes as per PR description.

**Comment by navidshaikh** [comment]
ping @bexelbie @LalatenduMohanty @budhrg , please review.

**Comment by bexelbie** [comment]
@navidshaikh this still needs docs and a changelog update

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,7 +1,7 @@
 # Changelog
 
-## Unreleased
 - Enhanced the developer instructions for developing the plugin in README
```
Comment (bexelbie) [inline CHANGELOG.md]: Can you add @budhrg to this line as it got ommitted

File: README.md [hunk 9]
```diff
@@ -32,6 +32,17 @@ The [Atomic Developer Bundle](https://github.com/projectatomic/adb-atomic-develo
 
 4. Begin using your host-based tools.
 
+## Exit codes
+
+Following table shows the exit codes from plugin and their meanings
```
Comment (bexelbie) [inline README.md]: The following table lists the plugin's exit codes and their meaning:

File: README.md [hunk 10]
```diff
@@ -32,6 +32,17 @@ The [Atomic Developer Bundle](https://github.com/projectatomic/adb-atomic-develo
...
+
+Exit Code Number | Meaning
+---------------  |-------------------------------------------------------------------------
+1                | Catch all for general errors / Wrong sub-command or option given
```
Comment (bexelbie) [inline README.md:42]: Add:

0 | No Error

File: README.md [hunk 11]
```diff
@@ -32,6 +32,17 @@ The [Atomic Developer Bundle](https://github.com/projectatomic/adb-atomic-develo
...
+3                | If vagrant box is not running and plugin is executed
```
Comment (bexelbie) [inline README.md]: The ADB is not running and must be before this command can succeed

File: CHANGELOG.md (hunk 8 above)
Comment (navidshaikh) [inline CHANGELOG.md]: I will update it when sending full list of changelog

**Comment by bexelbie** [comment]
`bundle exec vagrant service-manger env` returns an error message about openshift not running and exits with error code 126.  It shouldn't

print_all_provider_info assumes all providers are loaded and shouldn't.

File: README.md (hunk 11 above)
Comment (navidshaikh) [inline README.md]: I'll vote for "Vagrant box" instead of "ADB"

**Comment by navidshaikh** [comment]
> bundle exec vagrant service-manger env returns an error message about openshift not running and exits with error code 126. It shouldn't

Filed an issue for this https://github.com/projectatomic/vagrant-service-manager/issues/75

File: README.md (hunk 11 above)
Comment (bexelbie) [inline README.md]: that is fine

**Comment by bexelbie** [comment]
LGTM

**Pull request merged**